hacks/glx/gltext.man
.TH XScreenSaver 1 "25-Jul-98" "X Version 11"
.SH NAME
gltext \- draws text spinning around in 3D
.SH SYNOPSIS
.B gltext
[\-\-display \fIhost:display.screen\fP] [\-\-window] [\-\-root]
[\-\-window\-id \fInumber\fP]
[\-\-visual \fIvisual\fP] [\-\-delay \fImicroseconds\fP]
[\-\-text \fIstring\fP]
[\-\-program \fIcommand\fP]
[\-\-wander] [\-\-no-wander]
[\-\-spin \fIaxes\fP]
[\-\-no-spin]
[\-\-front] [\-\-no\-front]
[\-\-wireframe]
[\-\-fps]
.SH DESCRIPTION
The \fIgltext\fP program draws some text spinning around in 3D, using
a font that appears to be made of solid tubes.  
.SH OPTIONS
.I gltext
accepts the following options:
.TP 8
.B \-\-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-\-root
Draw on the root window.
.TP 8
.B \-\-window\-id \fInumber\fP
Draw on the specified window.
.TP 8
.B \-\-install
Install a private colormap for the window.
.TP 8
.B \-\-visual \fIvisual\fP\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-\-text \fIstring\fP
The text to display.  This may contain newlines, but it shouldn't be
very long.  The default is to display the machine name and OS version.

This may also be a format string acceptable to
.BR date (1)
and
.BR strftime (3) ,
in which case, it will be updated once a second.  So to make this
program display a spinning digital clock, you could do this:
.nf
.sp
	gltext -text "%A%n%d %b %Y%n%l:%M:%S %p"
.sp
.fi
To include a literal `%', you must double it: `%%'.

See the man page for
.BR strftime (3)
for more details.
.TP 8
.B \-\-program \fIcommand\fP
The given program is run, and its output is displayed.
If specified, this overrides \fI\-\-text\fP.
The output of this program will be repeatedely displayed, with new
pages of text shifting in every few seconds.  Lines should be relatively
short.  You might try:
.nf
.sp
	-program 'xscreensaver-text --cols 20'
.sp
.fi
.TP 8
.B \-\-maxlines
Set the number of lines of text to display.  By default, 
.I gltext
will print 8 lines of text at a time.  Use this option to increase or
decrease that number.  Be aware that 
.I gltext
is designed to work with a fairly small amount of text, so setting this
value too high might result in slow performance or strange behaviour
stemming from buffer overflows.  Increase at your own risk.
.TP 8
.B \-\-mono
Display the text in a monospace font.  Default is a variable-width font.
.TP 8
.B \-\-no\-mono
Display the text in a variable-width font.  This is the default.  
.TP 8
.B \-\-wander
Move the text around the screen.  This is the default.
.TP 8
.B \-\-no\-wander
Keep the text centered on the screen.
.TP 8
.B \-\-wander\-speed
Sets the speed at which the text wanders around.  Default is 0.02.  
.TP 8
.B \-\-spin
Which axes around which the text should spin.  The default is "XYZ",
meaning rotate it freely in space.  "\fB\-\-spin Z\fP" would rotate the
text in the plane of the screen while not rotating it into or out
of the screen; etc.
.TP 8
.B \-\-no\-spin
Don't spin the text at all: the same as \fB\-\-spin ""\fP.
.TP 8
.B \-\-front
When spinning, never spin all the way around or upside down:
always face mostly forward so that the text is easily readable.
.TP 8
.B \-\-no\-front
Allow spins to go all the way around or upside down.  This is the default.
.TP 8
.B \-\-wireframe
Render in wireframe instead of solid.
.TP 8
.B \-\-fps
Display the current frame rate, CPU load, and polygon count.
.TP 8
.B \-\-scale
Sets the scale at which the text is rendered.  Bigger values will result
in bigger text; smaller values will result in smaller text.  The default
value is 0.01.  
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.TP 8
.B XSCREENSAVER_WINDOW
The window ID to use with \fI\-\-root\fP.
.SH SEE ALSO
.BR X (1),
.BR xscreensaver (1)
.BR xscreensaver\-text (MANSUFFIX)
.SH COPYRIGHT
Copyright \(co 2001-2014 by Jamie Zawinski.
Permission to use, copy, modify, distribute, and sell this software and
its documentation for any purpose is hereby granted without fee,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation.  No representations are made about the
suitability of this software for any purpose.  It is provided "as is"
without express or implied warranty.
.SH AUTHOR
Jamie Zawinski <jwz@jwz.org>


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


hacks/glx/gltext.c
/* gltext, Copyright (c) 2001-2021 Jamie Zawinski <jwz@jwz.orgq2
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation.  No representations are made about the suitability of this
 * software for any purpose.  It is provided "as is" without express or 
 * implied warranty.
 */

#define DEFAULTS	"*delay:	20000        \n" \
			"*showFPS:      False        \n" \
			"*wireframe:    False        \n" \
			"*usePty:       False        \n" \

# define release_text 0
#define SMOOTH_TUBE       /* whether to have smooth or faceted tubes */

#ifdef SMOOTH_TUBE
# define TUBE_FACES  12   /* how densely to render tubes */
#else
# define TUBE_FACES  8
#endif


#include "xlockmore.h"
#include "colors.h"
#include "tube.h"
#include "sphere.h"
#include "rotator.h"
#include "gltrackball.h"
#include "textclient.h"
#include "utf8wc.h"

#include <ctype.h>

#ifdef USE_GL /* whole file */

#define DEF_TEXT          "(default)"
#define DEF_PROGRAM       "xscreensaver-text --date --cols 20 --lines 3"
#define DEF_SCALE_FACTOR  "0.01"
#define DEF_WANDER_SPEED  "0.02"
#define DEF_MAX_LINES     "8"
#define DEF_SPIN          "XYZ"
#define DEF_WANDER        "True"
#define DEF_FACE_FRONT    "True"
#define DEF_USE_MONOSPACE "False"

#include "glutstroke.h"
#include "glut_roman.h"
#include "glut_mroman.h"
#define GLUT_VARI_FONT (&glutStrokeRoman)
#define GLUT_MONO_FONT (&glutStrokeMonoRoman)
#define GLUT_FONT ((use_monospace) ? GLUT_MONO_FONT : GLUT_VARI_FONT)


typedef struct {
  GLXContext *glx_context;
  rotator *rot, *rot2;
  trackball_state *trackball;
  Bool button_down_p;
  Bool spinx, spiny, spinz;

  GLuint text_list;

  int ncolors;
  XColor *colors;
  int ccolor;

  char *text;
  int reload;

  time_t last_update;
  text_data *tc;

} text_configuration;

static text_configuration *tps = NULL;

static char   *text_fmt;
static char   *program_str;
static float  scale_factor;
static int    max_no_lines;
static float  wander_speed;
static char   *do_spin;
static Bool   do_wander;
static Bool   face_front_p;
static Bool   use_monospace;

static XrmOptionDescRec opts[] = {
  { "-text",         ".text",         XrmoptionSepArg, 0 },
  { "-program",      ".program",      XrmoptionSepArg, 0 },
  { "-scale",        ".scaleFactor",  XrmoptionSepArg, 0 },
  { "-maxlines",     ".maxLines",     XrmoptionSepArg, 0 },
  { "-wander-speed", ".wanderSpeed",  XrmoptionSepArg, 0 },
  { "-spin",         ".spin",         XrmoptionSepArg, 0 },
  { "+spin",         ".spin",         XrmoptionNoArg, "" },
  { "-wander",       ".wander",       XrmoptionNoArg, "True" },
  { "+wander",       ".wander",       XrmoptionNoArg, "False" },
  { "-front",        ".faceFront",    XrmoptionNoArg, "True" },
  { "+front",        ".faceFront",    XrmoptionNoArg, "False" },
  { "-mono",         ".useMonoSpace", XrmoptionNoArg, "True" },
  { "+mono",         ".useMonoSpace", XrmoptionNoArg, "False" }
};

static argtype vars[] = {
  {&text_fmt,      "text",         "Text",         DEF_TEXT,          t_String},
  /* This happens to be what utils/textclient.c reads */
  {&program_str,   "program",      "Program",      DEF_PROGRAM,       t_String},
  {&do_spin,       "spin",         "Spin",         DEF_SPIN,          t_String},
  {&scale_factor,  "scaleFactor",  "ScaleFactor",  DEF_SCALE_FACTOR,  t_Float},
  {&max_no_lines,  "maxLines",     "MaxLines",     DEF_MAX_LINES,     t_Int},
  {&wander_speed,  "wanderSpeed",  "WanderSpeed",  DEF_WANDER_SPEED,  t_Float},
  {&do_wander,     "wander",       "Wander",       DEF_WANDER,        t_Bool},
  {&face_front_p,  "faceFront",    "FaceFront",    DEF_FACE_FRONT,    t_Bool},
  {&use_monospace, "useMonoSpace", "UseMonoSpace", DEF_USE_MONOSPACE, t_Bool},
};

ENTRYPOINT ModeSpecOpt text_opts = {countof(opts), opts, countof(vars), vars, NULL};


/* Window management, etc
 */
ENTRYPOINT void
reshape_text (ModeInfo *mi, int width, int height)
{
  GLfloat h = (GLfloat) height / (GLfloat) width;
  int y = 0;

  if (width > height * 5) {   /* tiny window: show middle */
    height = width * 9/16;
    y = -height/2;
    h = height / (GLfloat) width;
  }

  glViewport (0, y, (GLint) width, (GLint) height);

  glMatrixMode(GL_PROJECTION);
  glLoadIdentity();
  gluPerspective (30.0, 1/h, 1.0, 100.0);

  glMatrixMode(GL_MODELVIEW);
  glLoadIdentity();
  gluLookAt( 0.0, 0.0, 30.0,
             0.0, 0.0, 0.0,
             0.0, 1.0, 0.0);

  {
    GLfloat s = (MI_WIDTH(mi) < MI_HEIGHT(mi)
                 ? (MI_WIDTH(mi) / (GLfloat) MI_HEIGHT(mi))
                 : 1);
    glScalef (s, s, s);
  }

  glClear(GL_COLOR_BUFFER_BIT);
}


static void
gl_init (ModeInfo *mi)
{
  text_configuration *tp = &tps[MI_SCREEN(mi)];
  int wire = MI_IS_WIREFRAME(mi);

  static const GLfloat pos[4] = {5.0, 5.0, 10.0, 1.0};

  if (!wire)
    {
      glLightfv(GL_LIGHT0, GL_POSITION, pos);
      glEnable(GL_CULL_FACE);
      glEnable(GL_LIGHTING);
      glEnable(GL_LIGHT0);
      glEnable(GL_DEPTH_TEST);
    }

  tp->text_list = glGenLists (1);
  glNewList (tp->text_list, GL_COMPILE);
  glEndList ();
}


static void
parse_text (ModeInfo *mi)
{
  text_configuration *tp = &tps[MI_SCREEN(mi)];
  char *old = tp->text;

  const char *tt =
    (text_fmt && *text_fmt && !!strcmp(text_fmt, "(default)")
     ? text_fmt : 0);
  const char *pr =
    (program_str && *program_str && !!strcmp(program_str, "(default)")
     ? program_str : 0);

  /* We used to do some "#ifdef HAVE_UNAME" stuff in here, but 
     "xscreensaver-text --date" does a much better job of that
     by reading random files from /etc/ and such.
   */

  if (tt && !strchr (tt, '%'))		/* Static text with no formatting */
    {
      tp->text = strdup (tt);
      tp->reload = 0;
    }
  else if (tt)				/* Format string */
    {
      time_t now = time ((time_t *) 0);
      struct tm *tm = localtime (&now);
      int L = strlen(text_fmt) + 100;
      tp->text = (char *) malloc (L);
      *tp->text = 0;
      strftime (tp->text, L-1, text_fmt, tm);
      if (!*tp->text)
        sprintf (tp->text, "strftime error:\n%s", text_fmt);
      tp->reload = 1;			/* Clock ticks every second */
    }
  else if (pr)
    {
      int max_lines = max_no_lines;
      char buf[4096];
      char *p = buf;
      int lines = 0;

      if (! tp->tc)
        /* This runs 'pr' because it reads the same "program" resource. */
        tp->tc = textclient_open (mi->dpy);

      while (p < buf + sizeof(buf) - 1 &&
             lines < max_lines)
        {
          int c = textclient_getc (tp->tc);
          if (c == '\n')
            lines++;
          if (c > 0)
            *p++ = (char) c;
          else
            break;
        }
      *p = 0;
      if (lines == 0 && buf[0])
        lines++;

      tp->text = strdup (buf);
      
      if (!*tp->text)
        tp->reload = 1;		/* No output, try again right away */
      else if (!strncmp (pr, "xscreensaver-text --date", 24))
        {
          /* If it's the default, and we have results, there's no need
             to ever reload. */
          tp->reload = 0;
        }
      else
        tp->reload = 7;		/* Linger a bit */
    }
  else
    abort();

  {
    /* The GLUT font only has ASCII characters. */
    char *s1 = utf8_to_latin1 (tp->text, True);
    free (tp->text);
    tp->text = s1;
  }

  /* If we had text before but got no text this time, hold on to the
     old one, to avoid flickering.
   */
  if (old && *old && !*tp->text)
    {
      free (tp->text);
      tp->text = old;
    }
  else if (old)
    free (old);
}


ENTRYPOINT Bool
text_handle_event (ModeInfo *mi, XEvent *event)
{
  text_configuration *tp = &tps[MI_SCREEN(mi)];

  if (gltrackball_event_handler (event, tp->trackball,
                                 MI_WIDTH (mi), MI_HEIGHT (mi),
                                 &tp->button_down_p))
    return True;

  return False;
}


ENTRYPOINT void 
init_text (ModeInfo *mi)
{
  text_configuration *tp;
  int i;

  MI_INIT (mi, tps);

  tp = &tps[MI_SCREEN(mi)];

  if ((tp->glx_context = init_GL(mi)) != NULL) {
    gl_init(mi);
    reshape_text (mi, MI_WIDTH(mi), MI_HEIGHT(mi));
  }

  {
    double spin_speed   = 0.5;
    double wander_speed = 0.02;
    double tilt_speed   = 0.03;
    double spin_accel   = 0.5;

    char *s = do_spin;
    while (*s)
      {
        if      (*s == 'x' || *s == 'X') tp->spinx = True;
        else if (*s == 'y' || *s == 'Y') tp->spiny = True;
        else if (*s == 'z' || *s == 'Z') tp->spinz = True;
        else if (*s == '0') ;
        else
          {
            fprintf (stderr,
         "%s: spin must contain only the characters X, Y, or Z (not \"%s\")\n",
                     progname, do_spin);
            exit (1);
          }
        s++;
      }

    tp->rot = make_rotator (tp->spinx ? spin_speed : 0,
                            tp->spiny ? spin_speed : 0,
                            tp->spinz ? spin_speed : 0,
                            spin_accel,
                            do_wander ? wander_speed : 0,
                            False);
    tp->rot2 = (face_front_p
                ? make_rotator (0, 0, 0, 0, tilt_speed, True)
                : 0);
    tp->trackball = gltrackball_init (False);
  }

  tp->ncolors = 255;
  tp->colors = (XColor *) calloc(tp->ncolors, sizeof(XColor));
  make_smooth_colormap (0, 0, 0,
                        tp->colors, &tp->ncolors,
                        False, 0, False);

  /* brighter colors, please... */
  for (i = 0; i < tp->ncolors; i++)
    {
      tp->colors[i].red   = (tp->colors[i].red   / 2) + 32767;
      tp->colors[i].green = (tp->colors[i].green / 2) + 32767;
      tp->colors[i].blue  = (tp->colors[i].blue  / 2) + 32767;
    }

  parse_text (mi);

}


static int
fill_character (GLUTstrokeFont font, int c, Bool wire, int *polysP)
{
  GLfloat tube_width = 10;

  const StrokeCharRec *ch;
  const StrokeRec *stroke;
  const CoordRec *coord;
  StrokeFontPtr fontinfo;
  int i, j;

  fontinfo = (StrokeFontPtr) font;

  if (c < 0 || c >= fontinfo->num_chars)
    return 0;
  ch = &(fontinfo->ch[c]);
  if (ch)
    {
      GLfloat lx=0, ly=0;
      for (i = ch->num_strokes, stroke = ch->stroke;
           i > 0; i--, stroke++) {
        for (j = stroke->num_coords, coord = stroke->coord;
             j > 0; j--, coord++)
          {
# ifdef SMOOTH_TUBE
            int smooth = True;
# else
            int smooth = False;
# endif

            if (j != stroke->num_coords)
              *polysP += tube (lx,       ly,       0,
                               coord->x, coord->y, 0,
                               tube_width,
                               tube_width * 0.15,
                               TUBE_FACES, smooth, False, wire);
            lx = coord->x;
            ly = coord->y;

            /* Put a sphere at the endpoint of every line segment.  Wasteful
               on curves like "0" but necessary on corners like "4". */
            if (! wire)
              {
                glPushMatrix();
                glTranslatef (lx, ly, 0);
                glScalef (tube_width, tube_width, tube_width);
                *polysP += unit_sphere (TUBE_FACES, TUBE_FACES, wire);
                glPopMatrix();
              }
          }
      }
      return (int) (ch->right + tube_width);
    }
  return 0;
}


static int
text_extents (const char *string, int *wP, int *hP)
{
  const char *s, *start;
  int line_height = GLUT_FONT->top - GLUT_FONT->bottom;
  int lines = 0;
  *wP = 0;
  *hP = 0;
  start = string;
  s = start;
  while (1)
    if (*s == '\n' || *s == 0)
      {
        int w = 0;
        while (start < s)
          {
            w += glutStrokeWidth(GLUT_FONT, *start);
            start++;
          }
        start = s+1;

        if (w > *wP) *wP = w;
        *hP += line_height;
        lines++;
        if (*s == 0) break;
        s++;
      }
    else
      s++;

  return lines;
}


static unsigned long
fill_string (const char *string, Bool wire)
{
  int polys = 0;
  const char *s, *start;
  int line_height = GLUT_FONT->top - GLUT_FONT->bottom;
  int off;
  GLfloat x = 0, y = 0;

  int ow, oh;
  text_extents (string, &ow, &oh);

  y = oh / 2 - line_height;

  start = string;
  s = start;
  while (1)
    if (*s == '\n' || *s == 0)
      {
        int line_w = 0;
        const char *s2;
        const char *lstart = start;
        const char *lend = s;

        /* strip off whitespace at beginning and end of line
           (since we're centering.) */
        while (lend > lstart && isspace(lend[-1]))
          lend--;
        while (lstart < lend && isspace(*lstart))
          lstart++;

        for (s2 = lstart; s2 < lend; s2++)
          line_w += glutStrokeWidth (GLUT_FONT, *s2);

        x = (-ow/2) + ((ow-line_w)/2);
        for (s2 = lstart; s2 < lend; s2++)
          {
            glPushMatrix();
            glTranslatef(x, y, 0);
            off = fill_character (GLUT_FONT, *s2, wire, &polys);
            x += off;
            glPopMatrix();
          }

        start = s+1;

        y -= line_height;
        if (*s == 0) break;
        s++;
      }
    else
      s++;
  return polys;
}


ENTRYPOINT void
draw_text (ModeInfo *mi)
{
  text_configuration *tp = &tps[MI_SCREEN(mi)];
  Display *dpy = MI_DISPLAY(mi);
  Window window = MI_WINDOW(mi);
  int wire = MI_IS_WIREFRAME(mi);

  GLfloat white[4] = {1.0, 1.0, 1.0, 1.0};
  GLfloat color[4] = {0.0, 0.0, 0.0, 1.0};

  if (!tp->glx_context)
    return;

  glXMakeCurrent(MI_DISPLAY(mi), MI_WINDOW(mi), *tp->glx_context);

  if (tp->reload)
    {
      if (time ((time_t *) 0) >= tp->last_update + tp->reload)
        {
          parse_text (mi);
          tp->last_update = time ((time_t *) 0);
        }
    }

  glShadeModel(GL_SMOOTH);

  glEnable(GL_DEPTH_TEST);
  glEnable(GL_NORMALIZE);
  glEnable(GL_CULL_FACE);


  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

  glPushMatrix ();
  glScalef(1.1, 1.1, 1.1);
  glRotatef(current_device_rotation(), 0, 0, 1);

  {
    double x, y, z;
    get_position (tp->rot, &x, &y, &z, !tp->button_down_p);
    glTranslatef((x - 0.5) * 8,
                 (y - 0.5) * 8,
                 (z - 0.5) * 8);

    gltrackball_rotate (tp->trackball);

    if (face_front_p)
      {
        double max = 90;
        get_position (tp->rot2, &x, &y, &z, !tp->button_down_p);
        if (tp->spinx) glRotatef (max/2 - x*max, 1, 0, 0);
        if (tp->spiny) glRotatef (max/2 - y*max, 0, 1, 0);
        if (tp->spinz) glRotatef (max/2 - z*max, 0, 0, 1);
      }
    else
      {
        get_rotation (tp->rot, &x, &y, &z, !tp->button_down_p);
        glRotatef (x * 360, 1, 0, 0);
        glRotatef (y * 360, 0, 1, 0);
        glRotatef (z * 360, 0, 0, 1);
      }
  }


  glColor4fv (white);

  color[0] = tp->colors[tp->ccolor].red   / 65536.0;
  color[1] = tp->colors[tp->ccolor].green / 65536.0;
  color[2] = tp->colors[tp->ccolor].blue  / 65536.0;
  tp->ccolor++;
  if (tp->ccolor >= tp->ncolors) tp->ccolor = 0;

  glMaterialfv (GL_FRONT, GL_AMBIENT_AND_DIFFUSE, color);

  glScalef(scale_factor, scale_factor, scale_factor);

  mi->polygon_count = fill_string(tp->text, wire);

  glPopMatrix ();

  if (mi->fps_p) do_fps (mi);
  glFinish();

  glXSwapBuffers(dpy, window);
}

ENTRYPOINT void
free_text(ModeInfo * mi)
{
  text_configuration *tp = &tps[MI_SCREEN(mi)];

  if (!tp->glx_context) return;
  glXMakeCurrent(MI_DISPLAY(mi), MI_WINDOW(mi), *tp->glx_context);

  if (tp->tc)
    textclient_close (tp->tc);
  if (tp->text) free (tp->text);
  if (tp->trackball) gltrackball_free (tp->trackball);
  if (tp->rot) free_rotator (tp->rot);
  if (tp->rot2) free_rotator (tp->rot2);
  if (tp->colors) free (tp->colors);
  if (glIsList(tp->text_list)) glDeleteLists(tp->text_list, 1);
}


XSCREENSAVER_MODULE_2 ("GLText", gltext, text)

#endif /* USE_GL */
